### PR TITLE
misc changes for thrust

### PR DESCRIPTION
**Pull request by pca006132**
1. Fix wasm build error.
2. Use std implementation where possible, so we can make thrust an optional dependency later (#138)

Problems:
1. We need some iterator library that is compatible with thrust in order to make thrust optional. The problem with https://github.com/dpellegr/ZipIterator is that it does not work with some thrust functions, but I'm not sure why, perhaps due to not implementing random access iterator? I haven't yet dig into it. And we need implementation for permutation iterator, transform iterator, etc.
2. Currently, the type of zip iterator has to be `thrust::zip_iterator<thrust::tuple<...>>` as they require something like `tail_type` in their zip iterator implementation which is not provided by `std::tuple`. This forces all functors to accept `thrust::tuple` and use thrust functions like `thrust::get` everywhere. We should be able to fix this by finding some other implementation of zip iterator.

### CODE REVIEW

File: thrust.diff [hunk 1]
```diff
@@ -0,0 +1,32 @@
+diff --git a/third_party/thrust/thrust/detail/type_traits.h b/third_party/thrust/thrust/detail/type_traits.h
+index d147f832..60b57654 100644
+--- a/third_party/thrust/thrust/detail/type_traits.h
++++ b/third_party/thrust/thrust/detail/type_traits.h
+@@ -144,11 +144,11 @@ template<typename T> struct has_trivial_constructor
+       is_pod<T>::value
+ #if THRUST_HOST_COMPILER == THRUST_HOST_COMPILER_MSVC || \
+     THRUST_HOST_COMPILER == THRUST_HOST_COMPILER_CLANG
+-      || __has_trivial_constructor(T)
++      || __is_trivially_constructible(T)
```
Comment (elalish) [inline thrust.diff:10]: Do we think thrust has a bug? Any idea what changed on the CI; can you repro locally?

**Comment by pca006132** [comment]
oh the patch file is wrong, will fix tmr

File: thrust.diff (hunk 1 above)
Comment (pca006132) [reply thrust.diff:10]: they use a deprecated API: https://github.com/NVIDIA/thrust/issues/1743

**Review by elalish — APPROVED** [review]
Thanks!

**Pull request merged**